top: Replace `umodule` with `module` everywhere.

jimmo opened the pull request:
Follow-up to #9018 -- this is the groundwork for doing the renaming of all built-in modules as well as renaming files and variables. I have always found the "u" prefix confusing, and although weak links provide a workaround there's still weirdness like the output of `help('modules')` etc. See #9018 for more but I strongly feel we should move in a direction of CPython compatibility and consistency.

Updated from an earlier version of this PR. This is now no longer a breaking change. It continues the decision in #11456 but makes the implementation of that more explicit by allowing modules to be defined as "extensible" or (default) "non-extensible" rather than inferring this from the u-prefix.

In summary:
- There is now no longer any `umodule` naming anywhere in the code (file names, qstrs, symbols, macros).
- A module can now be registered either as `MP_REGISTER_MODULE` or `MP_REGISTER_EXTENSIBLE_MODULE`. An extensible module can optionally be overridden from the filesystem, whereas a regular one will always bypass the filesystem. (An extensible module is what would have previously had a `u` prefix, e.g. `uos`).
- In order to retain backwards compatibility, importing an extensible module with it's `u` prefix will still work (this is kind of the inverse of weak links and is unconditionally enabled), and will bypass the filesystem. New code should prefer to use the `sys.path` mechanism introduced in #11456 instead (and hopefully we can deprecate the special u-handling later).
- `sys.path` is now assignable (using the same "mutable sys" mechansim as `sys.ps1` etc).

Reducing the size of the QSTRs has a small code size decrease (~100 bytes), but this is offset by the extra u-prefix handling and the split built-in modules maps, so there's a net negative code size increase (+12 bytes on PYBV11, +84 bytes on bare-arm). We could make the u-prefix handling conditional, but previously the minimal ports only supported "import umodule" so we should keep that working. However, this PR also removes the `timeq` module which gives a net saving around -850 bytes on PYBV11.

_This work was funded through GitHub Sponsors._

jimmo committed:
all: Rename MP_QSTR_umodule to MP_QSTR_module everywhere.

This renames the builtin-modules, such that help('modules') and printing
the module object will show "module" rather than "umodule".

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/builtinimport: Remove weak links.

In order to keep "import umodule" working, the existing mechanism is
replaced with a simple fallback to drop the "u".

This makes importing of built-ins no longer touch the filesystem, which
makes a typical built-in import take ~0.15ms rather than 3-5ms.

(Weak links were added in c14a81662c1df812c0c6b4299f97966302f16477)

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
all: Rename mp_umodule*, mp_module_umodule* to remove the "u" prefix.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
all: Rename mod_umodule*, ^umodule* to remove the "u" prefix.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
all: Rename UMODULE to MODULE in preprocessor/Makefile vars.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
all: Rename *umodule*.h to remove the "u" prefix.

This work was funded through GitHub Sponsors.

Also updates #includes.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
all: Rename *umodule*.c to remove the "u" prefix.

Updates any includes, and references from Makefiles/CMake.

This essentially reverts what was done long ago in commit
136b5cbd7669e8318f8455fc2706da97a5b7994c

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests: Replace umodule with module everywhere.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

robert-hh commented:
@jimmo 

> due to a limitation of the bash scripts I used to automate most of this. I will fix it up of course if we decide to merge.

Looking at the massive number files that have been changed this looks like needing a lot of changes.
Since I keep two ports alive until they get eventually merged (at least one of them), is is possible to provide the bash script you used for changing the sources?

jimmo committed:
examples/natmod: Rename umodule to module.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
I have updated the tests to translate

```
try:
  import foo
except:
  import foo
```

to just `import foo`.

> Since I keep two ports alive until they get eventually merged (at least one of them), is is possible to provide the bash script you used for changing the sources?

@robert-hh ... I did this as a series of (many!) one-liners. There was a bit of trial and error and some manual fixups.

Almost all of the changes are outside the ports, but in terms of rebasing a port, the one-liner below should get you a long way though if you run this on your branch before doing the rebase.

[umod.txt](https://github.com/micropython/micropython/files/9374518/umod.txt)

```
cd ports/samd
# See umod.txt attached to this comment.
cat ~/mpy/umod.txt | while read mod; do git grep -l $mod | xargs sed -i -r "s/(^|[^a-z])$mod($|[^a-z])/\1${mod:1}\2/g"; done
# ignore the no input files messages
```

robert-hh commented:
Thanks. I will try that when the time comes.

jimmo committed:
all: Replace all uses of umodule in Python code.

Applies to drivers/examples/extmod/port-modules/tools.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Now that #11456 is merged I have rebased & updated this.

For the time being I have not added a replacement for weak links. We need a way to annotate built-in modules as extensible-from-the-filesystem or not, and then match non-extensible ones at the start of import and extensible at the end. Also a mechanism to allow backwards-compatible `import ufoo`.

jimmo committed:
py/makemoduledefs.py: Add a way to register extensible built-in modules.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
all: Use MP_REGISTER_EXTENSIBLE_MODULE for overrideable built-ins.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Added a mechanism to replace the former weak link system and make explicit the behavior defined in https://github.com/micropython/micropython/pull/11456 about extensible vs non-extensible modules. Updated the PR description with details.

jimmo committed:
tests/run-multitests.py: Don't allow imports from the cwd.

Make tests run in an isolated environment (i.e. `import io` would
otherwise get the `tests/io` directory).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/run-perfbench.py: Don't allow imports from the cwd.

Make tests run in an isolated environment (i.e. `import io` would
otherwise get the `tests/io` directory).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/run-natmodtests.py: Don't allow imports from the cwd.

Make tests run in an isolated environment (i.e. `import io` would
otherwise get the `tests/io` directory).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/index: Update docs after umodule rename.

- Update guide for extending built-in modules.
- Remove any last trace of umodule in other docs.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
OK all tests passing at last!

@dlech @stinos @andrewleech I think this should alleviate the concerns from #9018. My intention with this PR is now that the only user-visible difference is the output of `help("modules")` and what you see if you print a module object (i.e. `<module 'os'>` rather than `<module 'uos'>`). All existing code should maintain the same behavior.

For people working in the codebase, the main change is the file and symbol renaming, but overall I believe this improves consistency and clarity.

andrewleech commented:
Having these as separate #defines is a great idea, though "extensible" sounds like a module you could setattr additional functionality onto which would be my preferred approach to extending if we could make an efficient mechanism to do so.

The functionality that's actually implemented is probably better described as "replaceable"?

The pattern you presented of clearing sys.path to get to a built-in, could you instead create a `builtins` module that only has a `__getattr__` function which always provides the C built-in function?

I guess this can be done in pure python with this current PR, wrapping up the sys.path pattern. It'd surely be more efficient in C though.

stinos commented on py/builtinimport.c:
```diff
@@ -378,28 +378,22 @@ STATIC mp_obj_t process_import_at_level(qstr full_mod_name, qstr level_mod_name,
         // relative to all the locations in sys.path.
         stat = stat_top_level(level_mod_name, &path);
 
         // TODO: If stat failed, now try extensible built-in modules.
```
The new code looks like it actually implements this, is this still TODO?

jimmo replied:
Yep, I updated the comment without removing the TODO :)

Used this opportunity to re-work the comments here a bit further.

stinos commented:
> The functionality that's actually implemented is probably better described as "replaceable"?

Technically it's indeed replacing, but on a higher level one will probably always implement it as an extension because the replacement normally contains all the original functionality as well. And since the notion 'extend' has been used here for quite a while now, it can stay as far as I'm concerned.


> I guess this can be done in pure python with this current PR, wrapping up the sys.path pattern. It'd surely be more efficient in C though.

You mean providing some function(s) which wrap the new boilerplate for forcing import of the builtin module, right? My first thought was also that we need that, but: does Python support that for star imports?

Apart from that boilerplate which is just somwehat annoying, I think this is all good.

jimmo commented:
> The pattern you presented of clearing sys.path to get to a built-in, could you instead create a `builtins` module that only has a `__getattr__` function which always provides the C built-in function?

Yep this was option (3) in #9018

If we handled this in the import mechanism, then it would work. But I think @stinos might be right that we couldn't emulate this from Python and still support wildcard imports.

> Apart from that boilerplate which is just somwehat annoying

It could be improved slightly (see https://github.com/micropython/micropython/pull/11456#issuecomment-1545949794) if we made it possible to assign to sys.path (which would be good for other reasons, so maybe we should do that anyway).

dpgeorge commented:
> if we made it possible to assign to sys.path (which would be good for other reasons, so maybe we should do that anyway).

Yes we should try to do that.

jimmo committed:
py/objmodule: Add a table of built-in modules with delegation.

This replaces the previous QSTR_null entry in the globals dict which could
leak out to Python (e.g. via iteration of mod.__dict__) and could lead to
crashes.

It results in smaller code size at the expense of turning a lookup into a
loop, but the list it is looping over likely only contains one or two
elements.

To allow a module to register its custom attr function it can use the new
`MP_REGISTER_MODULE_DELEGATION` macro.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/modsys: Allow sys.path to be assigned to.

Previously sys.path could be modified by append/pop or slice assignment.

This allows `sys.path = [...]`, which can be simpler in many cases, but
also improves CPython compatibility.

It also allows sys.path to be set to a tuple which means that you can
clear sys.path (e.g. temporarily) with no allocations.

This also makes sys.path (and sys.argv for consistency) able to be disabled
via mpconfig. The unix port (and upytesthelper) require them, so they
explicitly verify that they're enabled.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> > if we made it possible to assign to sys.path (which would be good for other reasons, so maybe we should do that anyway).
> 
> Yes we should try to do that.

OK I did that.

Added four more commits which:
 - Fixes #11466 with a different way of registering delegation handlers for built-in modules. (Diff is -32 bytes on PYBV11)
 - Makes sys.path use delegation (Diff is +36 bytes on PYBV11)
 - Updates builtin_ext.py test to use this (so that it results in no allocations)
 - Updates docs accordingly.

jimmo committed:
py/mpconfig: Enable module delegation if sys needs it.

Otherwise you can get into the confusing state where e.g. sys.ps1 is
enabled in config (via `MICROPY_PY_SYS_PS1_PS2`) but still doesn't actually
get enabled.

Also verify that the required delegation options are enabled in modsys.c.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge commented on py/mpconfig.h:
```diff
@@ -1421,6 +1421,11 @@ typedef double mp_float_t;
 
+// Whether to allow sys.path to be assigned to (rather than in-place updated)
+#ifndef MICROPY_PY_ALLOW_SYS_PATH_ASSIGNMENT
+#define MICROPY_PY_ALLOW_SYS_PATH_ASSIGNMENT (MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_EXTRA_FEATURES)
```
IMO this should always be enabled and not even have a config option.  If a port enables the sys module it gets a mutable sys.path.  Otherwise it's going to get a bit hairy trying to support ports that do not have a mutable version (basically you can't rely on the feature).

Actually, it might be worth making sys.path configurable via `MICROPY_SYS_PATH`.  Then if it is there it must be mutable.  That could be for ports that need sys but for where sys.path doesn't make sense.  But even then, all micropython-lib code should still assume that (a) sys exists, (b) sys.path is mutable.

jimmo replied:
Done - `sys.path` is now configurable via `MICROPY_SYS_PATH` and always assignable.

Also made sys.argv configurable too (even though it's just a regular read-only attribute). (Feels like someone who wants to disable sys.path is probably not interested in sys.argv).

dpgeorge commented on py/obj.h:
```diff
@@ -436,6 +436,8 @@ typedef struct _mp_rom_obj_t { mp_const_obj_t o; } mp_rom_obj_t;
 // As above, but allow this module to be extended from the filesystem.
 #define MP_REGISTER_EXTENSIBLE_MODULE(module_name, obj_module)
 
+// TODO
```
TODO?

dpgeorge commented:
The modules `uctypes`, `utimeq` and `ucryptolib` are "special" because they don't correspond to CPython (or other) modules (except for `uctypes` but it's very different to `ctypes` so should be considered independent).  So I don't think any of these should be renamed to remove their u prefix.

(And maybe we can remove `utimeq` because it's not really needed anymore, and I don't think there's much point in maintaining it.)

jimmo committed:
extmod/modtimeq: Remove timeq module.

This is a MicroPython-specific module that existed to support the old
version of uasyncio.  It's undocumented and not enabled on all ports and
takes up code size unnecessarily.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> The modules `uctypes`, `utimeq` and `ucryptolib` are "special" because they don't correspond to CPython (or other) modules (except for `uctypes` but it's very different to `ctypes` so should be considered independent). So I don't think any of these should be renamed to remove their u prefix.
> 
> (And maybe we can remove `utimeq` because it's not really needed anymore, and I don't think there's much point in maintaining it.)

The logic should be that the `u` prefix exists only to say "this module is _different_ to an equivalent CPython standard library module of the same name". If a module is similar/subset then it should get the exact CPython name. `cryptolib` is like `machine` or `network` in that it is a MicroPython module that has no CPython equivalent (abandoned PyPI project of the same name notwithstanding).

(By the same logic we should rename `urequests`, `uasyncio`, and `utarfile` as these are all subsets rather than different).

`cryptolib` was effectively renamed a while ago (it's already in the docs as cryptolib).

I have added a commit to remove `timeq`. I didn't realise this was only enabled on stm32/renesas+esp32+unix/windows, so this seems like a reasonable thing to do.

Agree r.e. `uctypes`. That was already unchanged in this PR.
There is one small quirk though with `uctypes`. Previously you could `import ctypes`, but that is no longer possible (although now you can do `import uuctypes`. :sob: ). I think you could argue that being able to `import ctypes` was a bug. If someone wants this to keep working they could write ctypes.py.

I would argue that we should therefore reconsider whether  `bluetooth`, `cryptolib`, `machine`, `uctypes`, `websocket` should be extensible.
- `machine` seems worthwhile because it's useful to be able to provide a stub version of hardware APIs (or even functional ones in some cases, e.g. embedded linux) on the Unix port.
- There could be existing code that uses `import ubluetooth`, `import cryptolib`, `import uwebsocket` so I think we need to keep them as extensible modules just so that keeps working. If we remove the reverse-weak-links feature in the future then I don't think we will need to keep these as extensible. Maybe we could just remove them from the list of extensible modules in the documentation.
- I don't think we should make `uctypes` extensible. That will solve the `import uuctypes` issue above.

jimmo committed:
py/objmodule: Workaround for MSVC with no module delegation.

When compiling mpy-cross, there is no `sys` module, and so there will
be no entries in the `mp_builtin_module_delegation_table`.

MSVC doesn't like this, so instead pretend as if the feature isn't
enabled at all.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge commented:
> Previously you could `import ctypes`, but that is no longer possible ... I think you could argue that being able to `import ctypes` was a bug

Yes I agree that was a bug.

dpgeorge commented on ports/cc3200/mods/moduos.c:
```diff
@@ -118,7 +118,7 @@ STATIC mp_obj_t os_urandom(mp_obj_t num) {
     return mp_obj_new_bytes_from_vstr(&vstr);
 }
-STATIC MP_DEFINE_CONST_FUN_OBJ_1(os_urandom_obj, os_urandom);
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(os_random_obj, os_random);
```
this urandom should not be renamed

dpgeorge commented on ports/nrf/modules/uos/moduos.c:
```diff
@@ -46,7 +46,7 @@
 /// \module os - basic "operating system" services
 ///
-/// The `os` module contains functions for filesystem access and `urandom`.
+/// The `os` module contains functions for filesystem access and `random`.
```
should stay

dpgeorge commented on ports/nrf/modules/uos/moduos.c:
```diff
@@ -110,7 +110,7 @@ STATIC mp_obj_t os_urandom(mp_obj_t num) {
     return mp_obj_new_bytes_from_vstr(&vstr);
 }
-STATIC MP_DEFINE_CONST_FUN_OBJ_1(os_urandom_obj, os_urandom);
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(os_random_obj, os_random);
```
all above should stay

dpgeorge commented on ports/nrf/modules/uos/moduos.c:
```diff
@@ -158,17 +158,17 @@ STATIC const mp_rom_map_elem_t os_module_globals_table[] = {
 
 #if MICROPY_HW_ENABLE_RNG
-    { MP_ROM_QSTR(MP_QSTR_random), MP_ROM_PTR(&os_urandom_obj) },
+    { MP_ROM_QSTR(MP_QSTR_random), MP_ROM_PTR(&os_random_obj) },
```
should stay

dpgeorge commented on extmod/moductypes.c:
```diff
@@ -718,6 +718,6 @@ const mp_obj_module_t mp_module_uctypes = {
 };
 
-MP_REGISTER_MODULE(MP_QSTR_uctypes, mp_module_uctypes);
+MP_REGISTER_EXTENSIBLE_MODULE(MP_QSTR_uctypes, mp_module_uctypes);
```
this shouldn't be extensible

jimmo commented:
> Yes I agree that was a bug.

OK, updated (plus the other nrf/cc3200 urandom fixes).

dpgeorge commented on py/objmodule.c:
```diff
@@ -171,12 +171,43 @@ STATIC const mp_rom_map_elem_t mp_builtin_module_table[] = {
+        }
+        #endif
+
+        if (extensible) {
```
Should this be `!extensible`?  Otherwise won't the code below check things like `upyb`?

jimmo replied:
The key is that it only looks at `mp_builtin_extensible_module_map` here. I've updated the comment to explain this better.

dpgeorge commented on ports/unix/main.c:
```diff
@@ -553,10 +561,8 @@ MP_NOINLINE int main_(int argc, char **argv) {
-    mp_obj_list_get(mp_sys_path, &path_num, &path_items);
-    path_items[0] = MP_OBJ_NEW_QSTR(MP_QSTR_);
+    mp_sys_path = mp_obj_new_list(0, NULL);
+    mp_obj_list_append(mp_sys_path, MP_OBJ_NEW_QSTR(MP_QSTR_));
```
Why not leave the `path_items` array as it was?  Ie:
```c
mp_sys_path = mp_obj_new_list(path_num, NULL);
mp_obj_t *path_items;
mp_obj_list_get(mp_sys_path, &path_num, &path_items);
...
```

jimmo replied:
This was to support having both assignable and non-assignable sys (which was later removed).

I've just rewritten this to do a single loop. It's simpler now, and fixes a small bug where before this PR it did:

```
$ env 'MICROPYPATH=:~/a:b' ./build-standard/micropython -c 'import sys; print(sys.path)'
['', '', '/home/jimmo/a', 'b']
```
now correctly does
```
$ env 'MICROPYPATH=:~/a:b' ./build-standard/micropython -c 'import sys; print(sys.path)'
['', '/home/jimmo/a', 'b']
```

dpgeorge replied:
OK

dpgeorge commented on py/modsys.c:
```diff
@@ -231,7 +239,6 @@ void mp_module_sys_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
     { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_sys) },
 
-    { MP_ROM_QSTR(MP_QSTR_path), MP_ROM_PTR(&MP_STATE_VM(mp_sys_path_obj)) },
     { MP_ROM_QSTR(MP_QSTR_argv), MP_ROM_PTR(&MP_STATE_VM(mp_sys_argv_obj)) },
```
This should probably be guarded by `MICROPY_PY_SYS_ARGV` now.

dpgeorge commented on tests/net_inet/test_tls_sites.py:
```diff
@@ -1,13 +1,8 @@
-    import ssl
-
-    # CPython only supports server_hostname with SSLContext
+import socket
```
instances of `_socket` need to be replaced with `socket` in this test

dpgeorge commented on docs/library/index.rst:
```diff
@@ -191,43 +191,55 @@ The following libraries are specific to the Zephyr port.
   sys.path = ()
   try:
-    from time import *
+    from os import *
```
shouldn't this stay as `time`?

dpgeorge commented:
I have finished a full and detailed review of this PR.  There are many changes done in the commits here, but a lot of them are simple renames, and most of those are in boilerplate code (eg defining modules and functions).

It's definitely great to see things like the tests simplified to just do `import time`, instead of the existing `try import utime as time; except ImportError: import time`.

Overall the changes here are a really good step forward.  It simplifies module naming, makes MicroPython behave more like CPython, still allows extensibility of built-ins.

dpgeorge commented:
Thanks @jimmo for all the hard work on this one.
